README
FFmpeg version 0.01
====================
FFmpeg (in this module, referred to here as FFmpeg-Perl) is a Perl
interface to the base project FFmpeg (referred to here as FFmpeg-C).
From the FFmpeg-C homepage:

FFmpeg-C is a complete solution to record, convert and stream audio and
video. It includes libavcodec, the leading audio/video codec library.
FFmpeg-C is developed under Linux, but it can compiled under most OSes,
including Windows.

FFmpeg-Perl currently only supports the functionality of the ffmpeg and
libavformat components  of the FFmpeg-C suite.  That is, functions
exist for extracting metadata from media streams and transforming one
media stream format to another, but no effort is (yet) made to port
HTTP broadcasting or playback functionality (provided by the ffserver
and ffplay components, respectively).

PREREQUISITES

Download and install the CVS version of FFmpeg.  The installation must
be configured with shared library support so libavformat and libavcodec
can be dynamically linked.  Make sure the install path of the shared
libraries is in the dynamic link path.  In Linux, this means locating
the libavformat.so and libavcodec.so files, and making sure their path
is specified in ld.so.conf.  Be sure to run "ldconfig" *twice* after
editing ld.so.conf.

FFmpeg is under active development.  This library was developed and 
tested on the FFmpeg CVS HEAD tag as of Thu Sep 30 01:19:14 PDT 2004.
FFmpeg was configured and installed using the options:

configure \
  --enable-pthreads \
  --enable-mp3lame \
  --enable-shared \
  --enable-vorbis

Note that this requires libmp3lame, libvorbis, and libpthreads to be
installed.  FFmpeg's "make test" failed for me with this configuration
because the regression test suite doesn't like having shared libraries
enabled.  If you experience problems with FFmpeg-perl or FFmpeg-C, try
recompiling without the shared library support to see if regression
tests pass.

INSTALLATION

To install this module type the following:

   # set $FFMPEG_ROOT environmental variable to /path/to/ffmpeg/source
   perl Makefile.PL
   make
   make test

It is critical that $FFMPEG_ROOT be set so that The Perl/C interface code
is able to correctly bind to ffmpeg functions.

Then as a user with install privileges:

   make install

DEPENDENCIES

This module requires these other modules and libraries:

   Image::Magick           ?.??
   Image::Magick::Iterator 0.01
   Image::PBMlib           1.05
   Time::Piece,            1.08

COPYRIGHT AND LICENCE

Copyright (c) 2003-2004 by Allen Day

This library is released under GPL, the Gnu Public License


Changes
0.01 Wed Apr 25 02:15:22 PDT 2004
	initial release

0.02
	- added more FFmpeg::Stream attributes (height, width, channels, etc)

	- FFmpeg::Stream made to be an uninstantiated base class for specific
	stream types: FFmpeg::Stream::Video, ::Audio, ::Data, and ::Unknown.

	- removed Inline::C dependency by reimplementing in XS (Dan Sully)

	- added capture_frame() method to FFmpeg::StreamGroup.  this introduces
	a dependency on Image::Magick::Iterator and Image::Magick.

	- changed Makefile.PL to read from $FFMPEG_ROOT environmental variable
	rather than a Makefile.PL argument.

0.03	- removed FFmpeg.PMS from distro (oops, should have been gone in 0.02)

	- added method for accessing header duration for RIFF files to
	FFmpeg::Stream::Video FIXME not done yet.

	- added input_url(), input_url_max_size(), and input_url_referrer() to
	allow customized downloading of files or their headers.  if
	input_url_max_size() is used, a second HEAD request is issued to get
	the real file size so FFMpeg::StreamGroup->duration() still reports
	the correct value.  this introduces dependency on LWP::UserAgent and
	HTTP::Request to the library.

0.04	- added force_format() method.
	- removed Time::Piece dependency.  duration() now returns fractional
	seconds for the duration of the stream or stream group.  call
	duration(format => 'HMS') to get a HH::MM::SS formatted string.  i
	may at some point add support for a 'Time::Piece' and other formats.
	- corrected a duration bug that reported incorrect durations for long
	streams (sprintf unsigned int instead of signed int)

###TODO###
fix audio file (mp3 anyway) duration reporting.  there may be
a bug in FFmpeg, a bug in Time::Piece, or a bug in mod_perl with
initializing malloc()ed memory.  i have only observed the bug running
under mod_perl, so i still need to do some isolation debugging.

FFmpeg::ImageFormat seems to have lost most of its formats (all except sgi,
gif, and yuv).  This seems to be related to the changes that were made to
capture_frames() requiring a .pnm extension on the tempfiles, and ffmpeg
automatically detecting what image format was desired... need to figure out
how to get the available image formats into perl again.  they are still
available as shown by running 'ffmpeg -formats'.


Makefile.PL
use ExtUtils::MakeMaker;
use Config;

$FFMPEG_ROOT = $ENV{FFMPEG_ROOT};
#if(!defined($FFMPEG_ROOT)){
#  die "you didn't define environmental variable \$FFMPEG_ROOT";
#} elsif(opendir(ZZ,$FFMPEG_ROOT)){
#  closedir(ZZ);
#} else {
#  die "couldn't opendir on \$FFMPEG_ROOT ('$FFMPEG_ROOT'): $!";
#}

#
# ... now we're ready to start the module build & install process
#

$LDDLFLAGS = '';
$LDDLFLAGS = '-all_load' if $^O eq 'darwin';
$LDDLFLAGS = '-L/usr/pkg/lib' if $^O eq 'netbsd';

#open(FFMPEGH, ">ffmpeg.h") or die $!;
#print FFMPEGH "#include \"$FFMPEG_ROOT/config.h\"\n";
#print FFMPEGH "#include \"$FFMPEG_ROOT/ffmpeg.c\"\n";
#close(FFMPEGH);

WriteMakefile(
      'NAME'          => 'FFmpeg',
      'VERSION_FROM'  => 'FFmpeg.pm',
      'DEFINE'        => '-O',
      'LDDLFLAGS'     => "$Config{'lddlflags'} $LDDLFLAGS",

      'PREREQ_PM'     => {
#no longer needed as of 0.04
#              Time::Piece   => 1.08,
#              Devel::Leak => 0,
              File::Temp => 0,
              HTTP::Request => 0,
              Image::Magick::Iterator   => 0.01,
              Image::PBMlib => 1.05, #1.06?
              LWP::UserAgent => 0,
      },

      ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
          (ABSTRACT_FROM  => 'FFmpeg.pm', # retrieve abstract from module
           AUTHOR         => 'Allen Day <allenday@ucla.edu>') : ()),

      'PERL_MALLOC_OK'  => 1,
      'LIBS'            => ["-lavcodec -lavformat -lm -lz -ldl -lmp3lame -logg -lvorbis -lvorbisenc"],
      'INC'             => "-I. -I$FFMPEG_ROOT/libavutil -I$FFMPEG_ROOT/libavcodec -I$FFMPEG_ROOT/libavformat -I/usr/include/ffmpeg -I/usr/include/ffmpeg/libavcodec -I/usr/include/ffmpeg/libavformat -I/usr/include/ffmpeg/libavutil",
);
